src/platform/input/game-pads.js

```diff
@@ -74,56 +77,480 @@ const MAPS = {
+    }
+
+    /**
+     * Get the value of one of the analogue axes of the pad.
+     *
+     * @param {number} axis - The axis to get the value of, use constants {@link PAD_L_STICK_X}, etc.
+     * @returns {number} The value of the axis between -1 and 1.
+     */
+    getAxis(axis) {
+        return this.axes[axis] && Math.abs(this.axes[axis]) > deadZone ? this.axes[axis] : 0;
+    }
+}
+
 /**
  * Input handler for accessing GamePad input.
+ *
+ * @augments EventHandler
  */
-class GamePads {
+class GamePads extends EventHandler {
     /**
      * Create a new GamePads instance.
      */
     constructor() {
+        super();
+
+        /**
+         * Whether gamepads are supported by this device.
+         *
+         * @type {boolean}
+         */
         this.gamepadsSupported = !!navigator.getGamepads || !!navigator.webkitGetGamepads;
 
+        /**
+         * The list of current gamepads.
+         *
+         * @type {GamePad[]}
+         */
         this.current = [];
+
+        /**
+         * The list of previous button states.
+         *
+         * @type {boolean[][]}
+         */
         this.previous = [];
 
-        this.deadZone = 0.25;
+        window.addEventListener('gamepadconnected', (event) => {
+            const pad = new GamePad(event.gamepad, this.getMap(event.gamepad));
+
+            while (this.current.findIndex(gp => gp.index === pad.index) !== -1) {
+                this.current.splice(this.current.findIndex(gp => gp.index === pad.index), 1);
+            }
+
+            this.current.push(pad);
+            this.fire(EVENT_GAMEPADCONNECTED, pad);
+        });
+
+        window.addEventListener('gamepaddisconnected', (event) => {
+            const padIndex = this.current.findIndex(gp => gp.index === event.gamepad.index);
+
+            if (padIndex !== -1) {
+                this.fire(EVENT_GAMEPADDISCONNECTED, this.current[padIndex]);
+                this.current.splice(padIndex, 1);
+            }
+        });
+    }
+
+    /**
+     * Fired when a gamepad is connected.
+     *
+     * @event GamePads#gamepadconnected
+     * @param {GamePad} gamepad - The gamepad that was just connected.
+     * @example
+     * var onPadConnected = function (pad) {
+     *     // Make the gamepad pulse.
+     * };
+     * app.keyboard.on("gamepadconnected", onPadConnected, this);
+     */
+
+    /**
+     * Fired when a gamepad is disconnected.
+     *
+     * @event GamePads#gamepaddisconnected
+     * @param {GamePad} gamepad - The gamepad that was just disconnected.
+     * @example
+     * var onPadDisconnected = function (pad) {
+     *     // Pause the game.
+     * };
+     * app.keyboard.on("gamepaddisconnected", onPadDisconnected, this);
+     */
+
+    set deadZone(value) {
+        deadZone = value;
+    }
+
+    /**
+     * @type {number} Threshold for axes to return values. Must be between 0 and 1.
+     * @ignore
+     */
+    get deadZone() {
+        return deadZone;
```

Comment:
Is this an over-engineered static variable now?

Reply:
As I didn't added a reference to the GamePads on the GamePad class I had to make this variable global to be able to access it. Added getter/setter for compatibility on it being GamePads variable at first

Reply:
It might make sense to give each GamePadButton a reference to GamePad, to receive the deadZone via `this.gamePad.deadZone`. But on the other hand, this is still not fine-grained enough.

For example clamping the steering wheel doesn't make sense, but other axes on it should be clamped. IMO there needs to be a system to enable clamping on a per-axis-base.

Reply:
Ideally the clamping code can be skipped/ignored entirely, if it's known that it's useless/harmful-to-precision (like for a steering wheel). As it is right now, I would need to set it to 0 to control e.g. a racing car in the full range, but then other players (lets say playing with a normal gamepad) would be affected by that setting aswell.

I don't see how a mapped axis button hurts this, you wouldn't assign the steering wheel axis as a button.

Reply:
A circular reference like this does not excite me at all .. it suggest it should be designed differently in most cases.

I'd suggest to step away from the code a bit and work on a a design of what this system should look like: what functionality is required and what should the public API for this to be. Ideally in comparison to what the current design is, to see where a backwards compatibility would change.

Reply:
To me the best option is to keep a global variable and then up to the user to make it 0 if they need more precision and use their own precision on the script. Fast, reliable and simple.

But question is if `GamePadButton`s should use `deadZone`, in current code it doesn't.

Reply:
@MushAsterion

There is no need for a `GamePadAxis`, you can give `GamePadButton` a reference to the `GamePad` and figure out if this particular button needs mapping from an axis based on `GamePad#map`. If so, the `GamePad#pressed` will tell so and that will also make "GamePadButton#wasPressed" work for an axis button.

So the extra information a `GamePadButton` requires is if it's mapped to an axis and then to which one in the form of a number/index.

@mvaligursky In this case I would suggest to deprecate `GamePads#deadZone`, as it makes least sense on `GamePads`. It would make more sense on individual GamePad's, but then it makes most sense on individual axes.

Reply:
> There is no need for a `GamePadAxis`

If you want a per-axis deadZone it's the easiest way and most understandable

> you can give `GamePadButton` a reference to the `GamePad` and figure out if this particular button needs mapping from an axis based on `GamePad#map`. If so, the `GamePad#pressed` will tell so and that will also make "GamePadButton#wasPressed" work for an axis button.

In fact the `GamePadButton#wasPressed` already works for an axis button from the getter, the only moment it's not working is if you don't fetch the most recent data. But again it introduces a circular reference.

Reply:
Can we consider this conversation resolved? Anything left to be decided here?